text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,243 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+- *root buildpack* - a special case of buildpack that is run with privileges
+
+Application developers can use root buildpacks to extend their build and/or run images.
+
+We introduce a boolean `privileged` key in the `[buildpack]` table of `buildpack.toml`, which is defined as follows:
+
+```
+[buildpack]
+privileged = <boolean (default=false)>
+```
+
+When `privileged` is set to `true`, the lifecycle will run this buildpack as the `root` user.
+
+For each root buildpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the buildpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively to snapshotting, a platform may store a new stack iamge to cache the changes. All of the captured changes will be included in a single layer produced as output from the buildpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+The buildpack can then exclude directories from the layer by writing to the `launch.toml` using a new `[[excludes]]` table, for example:
+
+```toml
+[[excludes]]
+paths = [ "/var" ]
+cache = true
+```
+
+* `paths` - a list of path globs to exclude from the run image
+* `cache` - if set to `true`, the paths will be stored in the cache
+
+The `[process]` and `[[slices]]` tables can be used as normal.
+
+The following constraint(s) will be enforced:
+* If a user attempts to create a buildpackage including both root buildpacks and non-root buildpacks, the process will fail.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Building an App with Additional Packages
+
+Given a root buildpack with ID `example/git`, the following configuration can be used with a Golang app:
+
+`project.toml`:
+```toml
+[[build.buildpacks]]
+id = "example/git"
+version = "1.0"
+
+[[build.buildpacks]]
+id = "example/golang"
+version = "1.0"
+```
+
+When the following command is run, the app source code will be loaded into the image for access by the root buildpack.
+
+`pack build sclevine/myapp`
+
+All root buildpacks will be sliced out of the list of buildpacks, and run before non-root buildpacks. This creates a version of the current builder with additional layers and an ephemeral run image with additional layers, then does a normal build phase without running the root buildpacks.
+
+After the build phase, *a new extend phase* will run the root buildpacks against the run-image to create an ephemeral base image that will be used by the export phase instead of the configured run-image.
+
+* The run image layers are persistented as part of the app image, and can be reused on subsequent builds.
+* The build image layers will be persisted as part of the cache, and can be reused on subsequent builds.
+* Any paths that are excluded using `launch.toml` and set to be cached will be stored in the cache, and can be reused on subsequent builds.
+
+## Upgrade an App
```

Comment:
While I understand the allure of a pack upgrade flow, I feel pretty strongly that this should just be accomplished by a complete rebuild. An "upgrade" that re-executes some but not all buildpacks seems arbitrary and confusing.

Reply:
The `pack upgrade` flow is just supposed to run the root buildpacks, and only on the run image. This does bring up a serious issue though: the app isn't accessible during a rebase. 

@jkutner I think root buildpacks will need something like a `/bin/rebase` executable and/or some way of storing metadata about the app. Otherwise rebase would require the original source code.

Alternatively, we could only give root buildpacks access to metadata in project.toml and not the app directory.

Reply:
Providing a `/bin/rebase` or restricting access to the app source seem to be workarounds to enable the "half a build" execution of pack upgrade, it seems much simpler to just do a complete rebuild. .

Reply:
> Alternatively, we could only give root buildpacks access to metadata in project.toml and not the app directory.

I think this is the way to go

Reply:
+1 to @jromero and @matthewmcnew . I think adding an `upgrade` would be a confusion for new consumers, who know they need to do something for their app, but aren't sure what. If there's an explicit need at a later point, great, but adding it preemptively to me sounds like the confusion risks outweigh the benefits.